### PR TITLE
[Driver][SYCL]Adding 'aoc -vpfp-relaxed' with -fintelfpga and -fp-model=fast

### PR DESCRIPTION
**Pull request by hchilama**
Adding '-vpfp-relaxed' to the aoc call with -fintelfpga and -fp-model=fast

### CODE REVIEW

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 1]
```diff
@@ -1410,6 +1410,14 @@ void SYCLToolChain::AddImpliedTargetArgs(const llvm::Triple &Triple,
                                           ":" + BackendOptName));
     }
   }
+  // only pass -vpfp-relaxed for aoc with -fintelfpga and fp-mode=fast
+  if (Args.hasArg(options::OPT_fintelfpga)) {
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/SYCL.cpp:1414]: This doesn't look to limit the option for `aoc` only.  an additional check will be needed to be sure we are targeting hardware.

File: clang/test/Driver/sycl-offload-intelfpga.cpp [hunk 2]
```diff
@@ -252,3 +252,10 @@
 // RUN:   %clangxx -### -target x86_64-unknown-linux-gnu -fsycl -fsycl-targets=spir64_fpga-unknown-unknown %s 2>&1 \
 // RUN:   | FileCheck -check-prefix=CHK-DISCARD-VALUE-NAMES %s
 // CHK-DISCARD-VALUE-NAMES-NOT: clang{{.*}} "-fsycl-is-device"{{.*}} "-discard-value-names"
+
+/// Check for vpfp-relaxed in the aoc call
+// RUN: %clangxx -### -fintelfpga -ffp-model=fast -Xshardware %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-FPGA-FPMODEL %s
+// RUN: %clang_cl -### -fintelfpga /clang:-ffp-model=fast -Xshardware %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-FPGA-FPMODEL %s
+// CHK-FPGA-FPMODEL: aoc{{.*}} "-dep-files={{.*}}" "-vpfp-relaxed"
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-offload-intelfpga.cpp:261]: Add check to make sure this isn't added for emulation (not setting `-Xshardware`)

File: clang/lib/Driver/ToolChains/SYCL.cpp (hunk 1 above)
Comment (hchilama) [reply clang/lib/Driver/ToolChains/SYCL.cpp:1414]: Thanks for the review Mike. Does this check ".IsFPGAHWMode()" works?

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 3]
```diff
@@ -1410,6 +1410,14 @@ void SYCLToolChain::AddImpliedTargetArgs(const llvm::Triple &Triple,
                                           ":" + BackendOptName));
     }
   }
+  // only pass -vpfp-relaxed for aoc with -fintelfpga and fp-mode=fast
+  if (Args.hasArg(options::OPT_fintelfpga) && getDriver().IsFPGAHWMode()
+      && Triple.getSubArch() == llvm::Triple::SPIRSubArch_fpga) {
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/SYCL.cpp:1415]:
```suggestion
  if (Args.hasArg(options::OPT_fintelfpga) && getDriver().IsFPGAHWMode() &&
      Triple.getSubArch() == llvm::Triple::SPIRSubArch_fpga) {
```

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 4]
```diff
@@ -1410,6 +1410,14 @@ void SYCLToolChain::AddImpliedTargetArgs(const llvm::Triple &Triple,
                                           ":" + BackendOptName));
     }
   }
+  // only pass -vpfp-relaxed for aoc with -fintelfpga and fp-mode=fast
```
Comment (mdtoguchi) [inline clang/lib/Driver/ToolChains/SYCL.cpp:1413]:
```suggestion
  // only pass -vpfp-relaxed for aoc with -fintelfpga and -fp-model=fast
```

File: clang/test/Driver/sycl-offload-intelfpga.cpp [hunk 5]
```diff
@@ -252,3 +252,19 @@
...
+// RUN: %clangxx -### -fintelfpga -ffp-model=fast %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-NO-HARDWARE %s
+// RUN: %clang_cl -### -fintelfpga -ffp-model=fast %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-NO-HARDWARE %s
+// RUN: %clangxx -### -fintelfpga -Xshardware %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-NO-HARDWARE %s
+// RUN: %clang_cl -### -fintelfpga -Xshardware %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-NO-HARDWARE %s
+// CHK-FPGA-FPMODEL: aoc{{.*}} "-dep-files={{.*}}" "-vpfp-relaxed"
+// CHK-NO-HARDWARE-NOT: aoc{{.*}} "-dep-files={{.*}}" "-vpfp-relaxed"
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-offload-intelfpga.cpp:270]:
```suggestion
// CHK-NO-HARDWARE-NOT: "-vpfp-relaxed"
```

File: clang/test/Driver/sycl-offload-intelfpga.cpp [hunk 6]
```diff
@@ -252,3 +252,19 @@
...
 // CHK-DISCARD-VALUE-NAMES-NOT: clang{{.*}} "-fsycl-is-device"{{.*}} "-discard-value-names"
+
+/// Check for vpfp-relaxed in the aoc call
+// RUN: %clangxx -### -fintelfpga -ffp-model=fast -Xshardware %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-FPGA-FPMODEL %s
+// RUN: %clang_cl -### -fintelfpga /clang:-ffp-model=fast -Xshardware %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-FPGA-FPMODEL %s
+// RUN: %clangxx -### -fintelfpga -ffp-model=fast %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-NO-HARDWARE %s
+// RUN: %clang_cl -### -fintelfpga -ffp-model=fast %s 2>&1 \
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-offload-intelfpga.cpp:263]:
```suggestion
// RUN: %clang_cl -### -fintelfpga /clang:-ffp-model=fast %s 2>&1 \
```

File: clang/test/Driver/sycl-offload-intelfpga.cpp (hunk 2 above)
Comment (hchilama) [reply clang/test/Driver/sycl-offload-intelfpga.cpp:269]: Done

File: clang/test/Driver/sycl-offload-intelfpga.cpp [hunk 7]
```diff
@@ -252,3 +252,19 @@
...
+// RUN: %clangxx -### -fintelfpga -ffp-model=fast %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-NO-HARDWARE %s
+// RUN: %clang_cl -### -fintelfpga /clang:-ffp-model=fast %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-NO-HARDWARE %s
+// RUN: %clangxx -### -fintelfpga -Xshardware %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-NO-HARDWARE %s
+// RUN: %clang_cl -### -fintelfpga -Xshardware %s 2>&1 \
+// RUN:  | FileCheck -check-prefix=CHK-NO-HARDWARE %s
+// CHK-FPGA-FPMODEL: aoc{{.*}} "-dep-files={{.*}}" "-vpfp-relaxed"
+// CHK-NO-HARDWARE-NOT: aoc{{.*}} "-vpfp-relaxed"
```
Comment (mdtoguchi) [inline clang/test/Driver/sycl-offload-intelfpga.cpp:270]:
```suggestion
// CHK-NO-HARDWARE-NOT: "-vpfp-relaxed"
```
having the `aoc` check in here would always cause a not match, due to the fact that `aoc` is only used with hardware enabled.

**Review by mdtoguchi — APPROVED** [review]

**Comment by hchilama** [comment]
@intel/llvm-gatekeepers , this is ready to merge, please take a look this at your convenience - Thanks.

**Pull request merged**